server/mqtt.go

```diff
@@ -2833,6 +2960,30 @@ func (as *mqttAccountSessionManager) transferRetainedToPerKeySubjectStream(log *
 	return nil
 }
 
+func (as *mqttAccountSessionManager) getCachedRetainedMsg(subject string) *mqttRetainedMsg {
+	v, ok := as.rmsCache.Load(subject)
+	if !ok {
+		return nil
+	}
+	rm := v.(mqttRetainedMsg)
+	if rm.expiresFromCache.Before(time.Now()) {
+		as.rmsCache.Delete(subject)
```

Comment:
If we never request that retained message again, then we will never delete the cache entry, so we maybe either need something that periodically fires to check for expired messages, or we just need to loop through all of them here and evacuate all dead entries.

Reply:
There is a long-running go routine that takes care of expiration, see `cleaupRetainedMessageCache`